Fluid rework: initial removal of FluidType

FiniteReality opened the pull request:
This is based on some of the notes we came up with here: https://hackmd.io/@neoforged/BkyGkyJGA

Expect bugs and broken behavior, like missing hooks - we should discuss these and figure out what hooks we can simplify or remove.

FiniteReality committed:
Initial removal of FluidType

Things are broken, which is to be expected.

FiniteReality committed:
Partially fix fluid pushing being too strong

There is still a bug where two different states of a fluid add together
for the purposes of fluid pushing. This needs to be fixed.

FiniteReality committed:
Add a system for indicating related fluids

Primarily aimed at physics calculations, could also be used for transfer mods

FiniteReality committed:
Fix lava pushing too much

Soaryn commented:
Just to log a thought: Most if not all of my fluids have no real reason to be in world, so just to reinforce the request of I'd like to be able to register a singular fluid, that is not intended to be placed or flow. I do think having a texture for both regardless is still useful information and that could very well be a client register, as there are cases where the flowing texture is used to show movement in pipes or similar. 

For fog, the quickest Idea I have is to have which fluid is the higher average of the screen, that is the color it chooses. If you move from say a red fluid to a blue fluid, interpolate that color change over the course of about 200ms or so (rough estimate). So instead of interpolating the colors between each other all the time, it would only change when the greater screen space fluid changes as well as over time rather than instant.

FiniteReality commented:
> Just to log a thought: Most if not all of my fluids have no real reason to be in world, so just to reinforce the request of I'd like to be able to register a singular fluid, that is not intended to be placed or flow.

I agree. While the Fluid itself will get FluidStates/etc which would enable placement in the world, there will almost certainly be a hook to enable in-world placement from things like buckets. This'll have to be documented so that e.g. AE2 formation planes don't try to place those fluids in the world either.

> I do think having a texture for both regardless is still useful information and that could very well be a client register, as there are cases where the flowing texture is used to show movement in pipes or similar.

Yeah; I'm thinking of setting up a "baked" liquid model which contains this information, so that we can remove most of the patches from the liquid block renderer while encouraging modded users to use models. There has been some pushback in this regard though previously, so I'm not sure.

> For fog, the quickest Idea I have is to have which fluid is the higher average of the screen, that is the color it chooses. If you move from say a red fluid to a blue fluid, interpolate that color change over the course of about 200ms or so (rough estimate). So instead of interpolating the colors between each other all the time, it would only change when the greater screen space fluid changes as well as over time rather than instant.

Agreed. I think what we'll end up doing is just having a target/current value, and we'll lerp between them by some factor every frame. This'll be easiest to implement and will likely result in the smallest patch surface.

FiniteReality committed:
Revert some liquid render changes, fix swimming

FiniteReality committed:
Move most Fluid patches back into IFluidExtension, set more properties

ApexModder commented on src/main/java/net/neoforged/neoforge/common/world/FluidBehaviour.java:
```diff
@@ -0,0 +1,378 @@
+/**
+ * A rough duplicate of {@link BlockBehaviour}, aimed at fluids.
+ */
+public abstract class FluidBehaviour implements FeatureElement {
```
Little note for future reference, if we are going to be allowing fluids to be disabled via feature flags the following will also need updating
- `FeatureElement.FILTERED_REGISTRIES` should include the fluid registry
  - This allows `RegistryLookup.filterFeatures` to filter out disabled fluids when used to lookup fluids
- `FluidStack` should have a `isFluidEnabled` method
  - Can easily be derived from `ItemStack.isItemEnabled`
- `BucketItem` should validate if the fluid being picked up/placed is enabled
  - Should also override `requiredFeatures` similar to `BlockItem` to ensure the item is disabled if its associated `content` is disabled
  - Similar override and checks should also be done for `LiquidBlock`
- fluid `SlotDisplay` types should be overriding `isEnabled` to validate if associated fluid(s) are enabled
  - Similar to `ItemStackSlotDisplay` and `ItemSlotDisplay`
- ~~`RecipeManager.isIngredientEnabled` should be updated to also check for disabled fluid ingredients~~
  - ~~This should maybe be expanded in future to giving custom ingredients something like a `isEnabled` override for better extensibility, rather than hardcoding support for fluid ingredients~~
  <sub>Edit: Fluid ingredients are not 'custom ingredients' like i thought they were, this would be left up to modders implementing their own recipe systems to filter out disabled fluid recipes during resource loading much like vanilla does</sub>
- Our fluid interaction system/registry might also need updating to only trigger interactions for enabled fluids

FiniteReality replied:
I copied this from BlockBehaviour, but I'm uncertain over whether we should make fluids actually support feature flags. I've left it in for now under the assumption that we'll want it though.

XFactHD commented:
> Yeah; I'm thinking of setting up a "baked" liquid model which contains this information, so that we can remove most of the patches from the liquid block renderer while encouraging modded users to use models. There has been some pushback in this regard though previously, so I'm not sure.

A fluid model is not feasible or at least does not provide any benefits in my eyes. There are far too many variations to realistically cache anything, so you'd end up recreating the `BakedQuad`s from scratch every time, which is almost certainly more expensive than just throwing the vertex data at a `VertexConsumer` directly.

Two things that stand out to me in general:
- The addition of the `FluidBehaviour` parameter to the `Fluid` constructor is a major break of vanilla bincompat and will cause a riot among multiloader devs (and for good reason)
- As far as I can tell, this doesn't seem to actually improve anything about the patches being almost impossible to maintain across MC version updates

FiniteReality committed:
Add custom fluid rendering, use Milk as test case

It seems there's an oddity with movement/pushing that I can't quite figure out...

FiniteReality commented:
> > Yeah; I'm thinking of setting up a "baked" liquid model which contains this information, so that we can remove most of the patches from the liquid block renderer while encouraging modded users to use models. There has been some pushback in this regard though previously, so I'm not sure.
> 
> A fluid model is not feasible or at least does not provide any benefits in my eyes. There are far too many variations to realistically cache anything, so you'd end up recreating the `BakedQuad`s from scratch every time, which is almost certainly more expensive than just throwing the vertex data at a `VertexConsumer` directly.

You're right. I thought over this some more and it's not really worth it. I instead added an interface that takes some cues from entity renderers as well as our current named render type system.

> Two things that stand out to me in general:
> 
> * The addition of the `FluidBehaviour` parameter to the `Fluid` constructor is a major break of vanilla bincompat and will cause a riot among multiloader devs (and for good reason)
> 
> * As far as I can tell, this doesn't seem to actually improve anything about the patches being almost impossible to maintain across MC version updates

We don't maintain binary compat with vanilla as is. As for multiloader devs, ultimately they're already going to have issues with all of these hooks being moved into Fluid. I think that any solution would make multi-loader compat invariably harder.

FiniteReality committed:
Fix comment

FiniteReality committed:
Avoid doing fluid pushing twice for living entities

XFactHD commented:
> We don't maintain binary compat with vanilla as is. As for multiloader devs, ultimately they're already going to have issues with all of these hooks being moved into Fluid. I think that any solution would make multi-loader compat invariably harder.

We do maintain vanilla binary compat as much as possible. There's a reason why our patches are riddled with overloads.
As for the hooks: they don't make ML compat any harder than before. You already had to extend or self-mixin your fluid impl in the NF module to override the `FluidType` getter and implementing the hooks is no different.

TelepathicGrunt commented on patches/net/minecraft/world/entity/LivingEntity.java.patch:
```diff
@@ -48,6 +56,12 @@
+     protected void checkFallDamage(double p_20990_, boolean p_20991_, BlockState p_20992_, BlockPos p_20993_) {
+-        if (!this.isInWater()) {
++        if (!this.isInWater() && !this.isInFluid()) { // Neo: only do fluid pushing if it wasn't previously calculated
+             this.updateInWaterStateAndDoWaterCurrentPushing();
```
Copying from discord:

I think the better solution is to patch updateInWaterStateAndDoWaterCurrentPushing to take a boolean to control whether to run the fluid pushing logic or not. Something like this. Especially as all this code really needs to do is update the fluid states, not run pushing
<img width="1330" height="760" alt="image" src="https://github.com/user-attachments/assets/f3d2499f-6526-40a0-bba9-8422e4411add" />
